File: polyscript-server/clear_text_action.ts
import { EditorModel } from "./editor_model";
import { ReplaceTextAction } from "./replace_text_action";
import { Selection } from "./selection";
import { Point } from "./point";

export class ClearTextAction extends ReplaceTextAction {  
    constructor(username) {
        super(username, [""])
    }
  
    static fromJson(json: any) {
      return new ClearTextAction(json["username"]);
    }
    
    execute(model: EditorModel) {
      this.model = model;
      this.selectedRange = this.initSelectionForDeletingRange(model);
  
      this.deleteSelectedText();
      this.updateUsersAfterDeleting();
      this.insertText();
      this.updateUsersAfterInserting();
    }
  
    initSelectionForDeletingRange(model: EditorModel) {
      var userSelection = model.users.find((user) => user.username == this.username).selection.copy();
  
      console.log(userSelection);

      if (userSelection.isEmpty) {
        if (userSelection.start.x == 0 && userSelection.start.y != 0) {
          return new Selection(
            userSelection.start.add(new Point(model.file.lines[userSelection.start.y - 1].length, -1)),
            userSelection.start,
          );
        } else if (userSelection.start.x != 0) {
          return new Selection(
            userSelection.start.add(new Point(-1, 0)),
            userSelection.start,
          );
        } else {
          return userSelection;
        }
      } else {
        return userSelection;
      }
    }
  
    
    toJson() {
      return JSON.stringify({"action": "clear_text", "username": this.username});
    }
  }
  

File: polyscript/lib/model/actions/clear_text_action.dart
import 'dart:convert';
import 'dart:math';

import 'package:polyscript/model/actions/replace_text_action.dart';
import 'package:polyscript/model/editor_model.dart';
import 'package:polyscript/model/user_model.dart';

import '../action_names.dart';

class ClearTextAction extends ReplaceTextAction {
  @override
  String get actionName => clearText;

  ClearTextAction(username) : super(username, [""]);

  static ClearTextAction fromJson(json) {
    return ClearTextAction(json["username"]);
  }

  @override
  void execute(EditorModel model) {
    this.model = model;
    selectedRange = initSelectionForDeletingRange(model);

    deleteSelectedText();
    updateUsersAfterDeleting();
    insertText();
    updateUsersAfterInserting();
  }

  Selection initSelectionForDeletingRange(EditorModel model) {
    var userSelection = model.users.firstWhere((user) => user.name == username).selection.copy();

    if (userSelection.isEmpty) {
      if (userSelection.start.x == 0 && userSelection.start.y != 0) {
        return Selection(
          userSelection.start + Point(model.file.lines[userSelection.start.y - 1].first.length, -1),
          userSelection.start,
        );
      } else if (userSelection.start.x != 0) {
        return Selection(
          userSelection.start - const Point(1, 0),
          userSelection.start,
        );
      } else {
        return userSelection;
      }
    } else {
      return userSelection;
    }
  }

  @override
  toJson() {
    return jsonEncode({"action": actionName, "username": username});
  }
}
